pkg/docker/config/config.go

```diff
@@ -0,0 +1,271 @@
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return "", errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return "", errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return registry, nil
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0600); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file
+func readJSONFile(path string) (dockerConfigFile, error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error creating/opening file %q", path)
+	}
+	defer file.Close()
+
+	var auths dockerConfigFile
+	fStat, err := os.Stat(path)
+	if err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error getting stat of file %q", path)
+	}
+	if fStat.Size() > 0 {
+		raw, err := ioutil.ReadFile(path)
+		if err != nil {
+			return dockerConfigFile{}, errors.Wrapf(err, "error reading file %q", path)
+		}
+
+		if err = json.Unmarshal(raw, &auths); err != nil {
+			return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+		}
+	} else {
+		newMap := make(map[string]dockerAuthConfig)
+		auths.AuthConfigs = newMap
+	}
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func addAuthToJSON(url, auth, path string) error {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return errors.Wrapf(err, "error creating/opening file %q", path)
+	}
+	defer file.Close()
+
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	enAuth := dockerAuthConfig{Auth: auth}
+	auths.AuthConfigs[url] = enAuth
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return nil
+}
+
+// encodeUserAndPass returns the base64 encoding of user:pass
+func encodeUserAndPass(username, password string) string {
+	return base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+}
+
+func getAuthFromCredHelper(credHelper, registry string) (string, string, error) {
```

Comment:
This has no caller any more. Actually the “set” and ”remove” functions probably need support for credential helpers as well.

Reply:
I believe I have to call this function in `GetAuthentication`? How exactly do credential helpers work? What should be passed in for `credHelper` in this function for example?

Reply:
Yes, each of get/set/remove should support helpers (so that after setting a credential, get returns that value, not some other one).

The documentation I’ve been pointed at is https://docs.docker.com/engine/reference/commandline/login/#credential-helper-protocol .

As for `credHelper`, see the code in the removed `getAuth`. I would expect the set/remove paths to be very similar.

I have never used the helpers either, so I’m not sure how to test the support. @dlorenc may be able to help.

Reply:
Thanks @mtrmac 
@dlorenc I guess my question is how should I make the code know when to store to `credHelpers` and when to `auths`?

Reply:
@umohnani8 `getAuth` was using (exclusively) the helper if the config file already contained a `CredHelpers[registry]` entry. Presumably the write paths should work the same way: if the user has configured the system to use a helper for a specific host name, use it (and only it); otherwise store the credentials in the config file.

Reply:
@mtrmac so the first entry to `credHelpers` will have to be done manually by the user by editing the `auth.json` file?
one more question - looking at https://github.com/GoogleCloudPlatform/docker-credential-gcr/blob/master/README.md there are 2 versions of cred helpers based on the version of Docker, which one are we supporting?

Reply:
> @mtrmac so the first entry to `credHelpers` will have to be done manually by the user by editing the `auth.json` file?

That would be good enough for me. Not sure about @dlorenc .

> there are 2 versions of cred helpers based on the version of Docker, 

The previous code used the `credHelpers` map.